### PR TITLE
Fixes #3017. View TextDirection returns incorrect size on a vertical direction instance with AutoSize as false.

### LINKED ISSUE #3017: View TextDirection returns incorrect size on a vertical direction instance with AutoSize as false.
By default the View base class initialize with the `TextDirection.LeftRight_TopBottom` and calculate the `Frame` with the horizontal direction, if no value is provided in the constructor. When instantiating a `View` by providing the `TextDirection` through the property, the horizontal direction is considered and calculated. When the direction property is set and as the `AutoSize` is false then the `Frame.Width` will persist. But it shouldn't because the user is instantiating with the vertical direction at creating.
So, the `TextDirection` should be undefined by default. All the derived class that only need to deal with one direction, then they should defined the value on the initial properties initialization, otherwise if no value is provided at initialization the `TextDirection.LeftRight_TopBottom` will be used at the `OnAdded` (if added to another sub-view) and at the `BeginInit`, to ensure a view with a null `SuperView` also initialize with a valid value.

```cs
		[Theory, AutoInitShutdown]
		[InlineData (true)]
		[InlineData (false)]
		public void View_Draw_Vertical_Simple_TextAlignments (bool autoSize)
		{
			var text = "Hello World";
			var height = 20;
			var lblLeft = new View (text) { Height = height, TextDirection = TextDirection.TopBottom_LeftRight, AutoSize = autoSize };
			var lblCenter = new View (text) { X = 2, Height = height, TextDirection = TextDirection.TopBottom_LeftRight, AutoSize = autoSize, VerticalTextAlignment = VerticalTextAlignment.Middle };
			var lblRight = new View (text) { X = 4, Height = height, TextDirection = TextDirection.TopBottom_LeftRight, AutoSize = autoSize, VerticalTextAlignment = VerticalTextAlignment.Bottom };
			var lblJust = new View (text) { X = 6, Height = height, TextDirection = TextDirection.TopBottom_LeftRight, AutoSize = autoSize, VerticalTextAlignment = VerticalTextAlignment.Justified };
			var frame = new FrameView () { Width = Dim.Fill (), Height = Dim.Fill () };

			frame.Add (lblLeft, lblCenter, lblRight, lblJust);
			Application.Top.Add (frame);
			Application.Begin (Application.Top);
			((FakeDriver)Application.Driver).SetBufferSize (9, height + 2);

			Assert.True (lblLeft.AutoSize == autoSize);
			Assert.True (lblCenter.AutoSize == autoSize);
			Assert.True (lblRight.AutoSize == autoSize);
			Assert.True (lblJust.AutoSize == autoSize);
			Assert.True (lblLeft.TextFormatter.AutoSize == autoSize);
			Assert.True (lblCenter.TextFormatter.AutoSize == autoSize);
			Assert.True (lblRight.TextFormatter.AutoSize == autoSize);
			Assert.True (lblJust.TextFormatter.AutoSize == autoSize);
			Assert.Equal (new Rect (0, 0, 1, height), lblLeft.Frame);
			Assert.Equal (new Rect (2, 0, 1, height), lblCenter.Frame);
			Assert.Equal (new Rect (4, 0, 1, height), lblRight.Frame);
			Assert.Equal (new Rect (6, 0, 1, height), lblJust.Frame);
			if (autoSize) {
				Assert.Equal (new Size (1, 11), lblLeft.TextFormatter.Size);
				Assert.Equal (new Size (1, 11), lblCenter.TextFormatter.Size);
				Assert.Equal (new Size (1, 11), lblRight.TextFormatter.Size);
			} else {
				Assert.Equal (new Size (1, height), lblLeft.TextFormatter.Size);
				Assert.Equal (new Size (1, height), lblCenter.TextFormatter.Size);
				Assert.Equal (new Size (1, height), lblRight.TextFormatter.Size);
			}
			Assert.Equal (new Size (1, height), lblJust.TextFormatter.Size);
			Assert.Equal (new Rect (0, 0, 9, height + 2), frame.Frame);

			var expected = @"
┌───────┐
│H     H│
│e     e│
│l     l│
│l     l│
│o H   o│
│  e    │
│W l    │
│o l    │
│r o    │
│l   H  │
│d W e  │
│  o l  │
│  r l  │
│  l o  │
│  d    │
│    W W│
│    o o│
│    r r│
│    l l│
│    d d│
└───────┘
";

			var pos = TestHelpers.AssertDriverContentsWithFrameAre (expected, output);
			Assert.Equal (new Rect (0, 0, 9, height + 2), pos);
		}

```

### PR DESCRIPTION
**Pull request by BDisp**
Fixes #3017 - This PR highlights and clarifies how the `LayoutStyle` differs from each other and how they are actually managed by the `LayoutSubViews` method, where the `SetRelativeLayout` method is only executed if the `View` is `LayoutStyle.Computed`. Faced with this reality, unit tests must also faithfully simulate exactly how the application behaves.
I also restored broken unit tests. 

## Pull Request checklist:

- [x] I've named my PR in the form of "Fixes #issue. Terse description."
- [x] My code follows the [style guidelines of Terminal.Gui](https://github.com/gui-cs/Terminal.Gui/blob/develop/.editorconfig) - if you use Visual Studio, hit `CTRL-K-D` to automatically reformat your files before committing.
- [x] My code follows the [Terminal.Gui library design guidelines](https://github.com/gui-cs/Terminal.Gui/blob/develop/CONTRIBUTING.md)
- [x] I ran `dotnet test` before commit
- [ ] I have made corresponding changes to the API documentation (using `///` style comments)
- [x] My changes generate no new warnings
- [x] I have checked my code and corrected any poor grammar or misspellings
- [x] I conducted basic QA to assure all features are working

### CODE REVIEW

File: Terminal.Gui/Text/TextFormatter.cs [hunk 1]
```diff
@@ -1046,7 +1046,7 @@ public static string RemoveHotKeySpecifier (string text, int hotPos, Rune hotKey
 		string _text = null;
 		TextAlignment _textAlignment;
 		VerticalTextAlignment _textVerticalAlignment;
-		TextDirection _textDirection;
+		TextDirection _textDirection = (TextDirection)(-1);
```
[the remark this answers is not in the document]
Comment (BDisp) [reply Terminal.Gui/Text/TextFormatter.cs:1049]: Nothing can prevent a user from setting an integer enum to -1, -2, -3, etc. In this case we are preventing -1. Done.

[the remark this answers is not in the document]
Comment (BDisp) [reply Terminal.Gui/Text/TextFormatter.cs:1049]: I improved this to avoids any invalid value in the commit https://github.com/gui-cs/Terminal.Gui/pull/3019/commits/644a78f86c7516c29807c471235606ab87954ea2.

[the remark this answers is not in the document]
Comment (tig) [reply Terminal.Gui/Text/TextFormatter.cs:1049]: I do not understand why TextDirection can't be set before Init? Why not just use `IsInitialized` to ensure it's not accessed?

[the remark this answers is not in the document]
Comment (BDisp) [reply Terminal.Gui/Text/TextFormatter.cs]: I changed this as it was before but using `IsInitialized`. I hope you like the solution.

File: UnitTests/Application/MouseTests.cs [hunk 2]
```diff
@@ -98,6 +98,11 @@ public void MouseCoordinatesTest_NoFrames (int offset, int clickX, int clickY, i
 		Application.Top.Y = pos.Y;
 		Application.Top.Width = size.Width;
 		Application.Top.Height = size.Height;
+		Assert.Equal (LayoutStyle.Absolute, Application.Top.LayoutStyle);
+		Assert.Equal (new Rect (0, 0, 80, 25), Application.Top.Frame);
+		Application.Top.LayoutStyle = LayoutStyle.Computed;
+		Application.Top.SetRelativeLayout (new Rect (Point.Empty, new Size (Application.Driver.Cols, Application.Driver.Rows)));
+		Assert.Equal (new Rect (offset, offset, 10, 10), Application.Top.Frame);
```
Comment (BDisp) [inline UnitTests/Application/MouseTests.cs:105]: With my new change this unit test fails and the right reason. `Application.Top.LayoutStyle` is always `LayoutStyle.Absolute`. Thus the Application.Top.Frame is always the same as `new Rect (Point.Empty, new Size (Application.Driver.Cols, Application.Driver.Rows))`, no matter the `Pos/Dim` are changed. The solution was to change to `LayoutStyle.Computed` and call the `SetRelativeLayout` with the `Driver.Cols/Driver.Rows` dimension.

File: Terminal.Gui/Views/CheckBox.cs [hunk 3]
```diff
@@ -89,6 +89,7 @@ void SetInitialProperties (string s, bool is_checked)
 			Checked = is_checked;
 			HotKeySpecifier = (Rune)'_';
 			CanFocus = true;
+			TextDirection = TextDirection.LeftRight_TopBottom;
```
[the remark this answers is not in the document]
Comment (BDisp) [reply Terminal.Gui/Views/CheckBox.cs:92]: I will remove it. Was to force the initialization but it isn't needed anymore. Thanks for see this.

[the remark this answers is not in the document]
Comment (BDisp) [reply Terminal.Gui/Views/CheckBox.cs]: Done.

File: Terminal.Gui/Views/Toplevel.cs [hunk 4]
```diff
@@ -193,7 +193,7 @@ public Toplevel () : base ()
 		void SetInitialProperties ()
 		{
 			ColorScheme = Colors.TopLevel;
-
+			TextDirection = TextDirection.LeftRight_TopBottom;
```
[the remark this answers is not in the document]
Comment (BDisp) [reply Terminal.Gui/Views/Toplevel.cs]: Removed. Thanks.

File: Terminal.Gui/Views/TimeField.cs [hunk 5]
```diff
@@ -70,6 +70,7 @@ public TimeField (TimeSpan time) : base (string.Empty)
 		void Initialize (TimeSpan time, bool isShort = false)
 		{
 			CultureInfo cultureInfo = CultureInfo.CurrentCulture;
+			LayoutStyle = LayoutStyle.Computed;
```
[the remark this answers is not in the document]
Comment (BDisp) [reply Terminal.Gui/Views/TimeField.cs]: Done. Thanks.

File: Terminal.Gui/Views/DateField.cs [hunk 6]
```diff
@@ -70,6 +70,7 @@ public DateField (DateTime date) : base ("")
 		void Initialize (DateTime date, bool isShort = false)
 		{
 			CultureInfo cultureInfo = CultureInfo.CurrentCulture;
+			LayoutStyle = LayoutStyle.Computed;
```
[the remark this answers is not in the document]
Comment (BDisp) [reply Terminal.Gui/Views/DateField.cs]: Done.

**Review by tig — APPROVED** [review]

**Comment by tig** [comment]
> @tig I'll give up trying to fix this because I' tired always resolving conflicts to nothing and become broken again, sorry.

Sorry you're tired and annoyed. 

One thing to consider here is `View.AutoSize` will likely go away if I can get `Dim.Auto` to work.

**Comment by BDisp** [comment]
But absolute layout is broken on the View constructor when you initialize with a frame and then OnResized change the frame again.

**Comment by BDisp** [comment]
> > @tig I'll give up trying to fix this because I' tired always resolving conflicts to nothing and become broken again, sorry.
> 
> Sorry you're tired and annoyed.
> 
Sorry for that. But I really had resolved many times conflicts on this and I think I missed some of my toughs. When you ask to me to stop on some pull request because we both are working on the same files, I understand and I stop from pushing anything more.

> One thing to consider here is `View.AutoSize` will likely go away if I can get `Dim.Auto` to work.

I think you are in the right way, but if this PR fix something related with the `TextDirection` at initialization and is good to merge, then you can continue work on your PR after merged. By the way, if you want to ensure that the `Dim.Auto` will work properly, my PR has unit tests that can be used to ensure that it really work, otherwise something is wrong.

**Comment by tig** [comment]
@bdisp - I forgot about this as I was working on the fix for #3127 (#3130).

I *think* I fixed all of this there!!!

I'll pour over this asap to see if I missed anything. As you review #3130, please let me know if you see anything I missed.

**Comment by tig** [comment]
We can close this now. I've carefully merged all of these excellent fixes (with appropriate changes) into #3130.  

I'll push to that PR soon you you can see!

File: UnitTests/View/ViewTests.cs [hunk 7]
```diff
@@ -1415,7 +1429,131 @@ public void Dispose_View ()
...
+";
+
+			var pos = TestHelpers.AssertDriverContentsWithFrameAre (expected, output);
+			Assert.Equal (new Rect (0, 0, width + 2, 6), pos);
+		}
 
+		[Theory, AutoInitShutdown]
+		[InlineData (true)]
+		[InlineData (false)]
+		public void View_Draw_Vertical_Simple_TextAlignments (bool autoSize)
```
Comment (BDisp) [inline UnitTests/View/ViewTests.cs:1491]: @tig this test you can't successful reproduce on your PR #3130.

Comment (tig) [reply UnitTests/View/ViewTests.cs:1491]: I don't know what you mean. I moved that test to #3130 and it passes.

image

Comment (BDisp) [reply UnitTests/View/ViewTests.cs:1491]: Ops I didn't pull, sorry. Confirmed thanks.

**Pull request closed without merging**